Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java

```diff
@@ -177,24 +187,41 @@ public boolean problemExists(int problemId) {
     }
 
     @Override
-    public void archiveProblemById(int problemId) {
-        Update update = new Update();
-        Query query = new Query(where("_id").is(problemId));
-        update.set("archived", true);
+    public void archiveProblem(Problem problem) {
+        Update update = update("archived", true);
+        Query query = query(where("_id").is(problem.getId()));
         mongoTemplate.updateFirst(query, update, ProblemData.class);
     }
 
     @Override
-    public void deleteProblemById(int problemId) {
-        Query query = new Query(where("_id").is(problemId));
-        mongoTemplate.remove(query, ProblemData.class);
+    public void deleteProblem(Problem problem) {
+        mongoTemplate.remove(query(where("_id").is(problem.getId())), ProblemData.class);
+        deleteFiles(problem);
     }
 
     @Override
     public void deleteAll() {
+        findAll().forEach(this::deleteFiles);
         mongoTemplate.dropCollection(ProblemData.class);
     }
 
+    private void deleteFiles(Problem problem) {
+        List<String> fileIds = new LinkedList<>();
+        var languageEnvs = problem.getLanguageEnvs();
+        if (languageEnvs != null) {
+            List<String> providedCodesFileIds = mapToList(languageEnvs.values(), LanguageEnv::getProvidedCodesFileId);
+            fileIds.addAll(providedCodesFileIds);
+        }
```

Comment:
the `languageEnvs` can never be null

in order to ensure this
modify Problem's `getLanguageEnvs` as follows
```java
public Map<String, LanguageEnv> getLanguageEnvs() {
        return requireNonNullElse(languageEnvs, emptyMap());;
}
```